File: genui/renderer/renderer.py
# -*- coding: utf-8 -*-
"""UI渲染器, 负责渲染和显示UI实例"""

from typing import Optional
from genui.core.ui_instance import UIInstance
from genui.adapters.base import AdapterBase
from genui.adapters.tkinter_adapter import TkinterAdapter
from genui.tools.registry import ToolRegistry
from genui.tools.executor import ToolExecutor
from genui.logger import get_logger

logger = get_logger(__name__)


class Renderer:
    """UI渲染器, 负责将UI实例渲染为可视化界面"""

    def __init__(
        self,
        adapter: Optional[AdapterBase] = None,
        tool_executor: Optional[ToolExecutor] = None
    ):
        """初始化渲染器

        Args:
            adapter: UI框架适配器, 如果为None则使用默认的TkinterAdapter
            tool_executor: Tool 执行器, 如果为 None 则创建默认实例
        """
        self.adapter = adapter or TkinterAdapter()
        self.tool_executor = tool_executor or ToolExecutor(ToolRegistry())
        logger.info(f"使用渲染适配器: {type(self.adapter).__name__}")

    def render(self, ui_instance: UIInstance) -> None:
        """渲染并显示UI实例

        Args:
            ui_instance: 要渲染的UI实例
        """
        # 创建窗口, 传递tool_executor
        window = self.adapter.create_window(ui_instance, self.tool_executor)

        # 运行事件循环
        self.adapter.run_event_loop(window)

    def render_to_window(self, ui_instance: UIInstance) -> object:
        """渲染UI实例到窗口但不运行事件循环

        Args:
            ui_instance: 要渲染的UI实例

        Returns:
            创建的窗口对象
        """
        return self.adapter.create_window(ui_instance, self.tool_executor)

    def run_window(self, window: object) -> None:
        """运行窗口的事件循环

        Args:
            window: 窗口对象
        """
        self.adapter.run_event_loop(window)

File: src/renderer/renderer.ts
// -*- coding: utf-8 -*-
// 渲染器, 连接UIAdapter与UIInstance, 负责渲染和运行事件循环

import type { UIAdapter } from '../adapters/base.js';
import type { UIInstance } from '../core/ui-instance.js';
import type { ToolExecutor } from '../tools/executor.js';
import { createLogger } from '../logger.js';

const logger = createLogger('Renderer');

/**
 * 渲染器类, 将UIInstance通过UIAdapter渲染到目标平台.
 *
 * 职责:
 * - 调用adapter.createWindow()创建窗口并渲染组件
 * - 调用adapter.runEventLoop()运行事件循环
 */
export class Renderer {
  /** UI适配器 */
  private readonly adapter: UIAdapter;

  /** 工具执行器(可选) */
  private readonly toolExecutor?: ToolExecutor;

  /**
   * 创建渲染器实例.
   *
   * @param adapter - UI适配器
   * @param toolExecutor - 可选的工具执行器
   */
  constructor(adapter: UIAdapter, toolExecutor?: ToolExecutor) {
    this.adapter = adapter;
    this.toolExecutor = toolExecutor;
    logger.info(`使用适配器: ${adapter.name}`);
  }

  /**
   * 渲染UI实例.
   * 先调用adapter.createWindow()创建窗口, 然后调用adapter.runEventLoop()运行事件循环.
   *
   * @param uiInstance - 要渲染的UI实例
   */
  async render(uiInstance: UIInstance): Promise<void> {
    logger.info(`渲染UI: ${uiInstance.title}`);
    await this.adapter.createWindow(uiInstance, this.toolExecutor);
    await this.adapter.runEventLoop();
  }
}
